### ISSUE TITLE
LoadImage's returned HANDLE must be released in ONE of THREE ways

### ISSUE THREAD

**Issue by AArnott** [microsoft/win32metadata#611]
`LoadImageW`/`LoadImageA` in the metadata currently return a `HANDLE` type, which would be closed with `CloseHandle`.
However [the docs](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadimagea#return-value) state that one of `DeleteObject`, `DeleteCursor` or `DeleteIcon` should be used to release the handle, and only if the `LR_SHARED` flag was not passed in. The metadata returning `HANDLE` means `CloseHandle` will *always* be called, which is the wrong method and the wrong conditions.

How can we model this in the metadata?

**Comment by marler8997** [comment]
Create a "LoadImageHandle" type with an attribute like "FreeFuncCandidates"?

**Comment by riverar** [comment]
This seems to be a problem specific to C# bindings and its use of SafeHandle. I don't think everyone else should shoulder the burden of a new type here.

Perhaps on the bindings side, functions with HANDLE return types should be generated with a non-SafeHandle overload?

**Comment by AArnott** [comment]
How is this C#-specific? Any projection that wants to help the user clean up resources (I think Rust is among these) will need to know how to do so. SafeHandle is just how .NET encapsulates this functionality.

**Comment by kennykerr** [comment]
Rust doesn't make use of `RAIIFree` at all. It would have to be absolutely perfect in terms of coverage and attribution for developers to feel confident in its use and even then its not necessarily applicable to all languages. But its an attribute, so it can be easily ignored. I would rather not introduce even more pseudo handle types as it muddies up the projection for languages that don't care about such things. If you must add more metadata for this, please just use an attribute that code generators can ignore if so desired.

**Comment by riverar** [comment]
My concern was regarding the suggestion to mutate the API signature to accommodate SafeHandle. You are right, the overall request is not C# specific, apologies for any confusion introduced.

We need to consider projections that may not want to do anything here.

I'm not pushing back on the idea of _annotating_ these APIs though. That will be great for bindings that wish to exert more control over object lifetime.

**Comment by mikebattista** [comment]
> Perhaps on the bindings side, functions with HANDLE return types should be generated with a non-SafeHandle overload?

Actually LoadImage already has this HANDLE-returning overload so I believe we're already doing this in C#/Win32. @AArnott is this correct?

**Comment by mikebattista** [comment]
So discussing with Andrew, he pointed out that we do always have an overload that matches the metadata, but HANDLE is wrong for this API since it implies CloseHandle which is wrong in all cases here. It should be changed to IntPtr at least to fix the crash that happens on disposal currently.

We could try to model this specific case in attributes, but it's a nontrivial conditional and I'd rather not go down that path for this one API. I'll tag this as a "special case" for tracking purposes but for now I think the change here is to change the return type in the metadata from HANDLE to IntPtr and expect developers to follow the documentation for the API.

**Comment by riverar** [comment]
> but HANDLE is wrong for this API since it implies CloseHandle which is wrong in all cases here.

I'm not aware of any Win32 guarantees or requirements that `HANDLE` objects be freeable via `CloseHandle`. So it's not clear why this is being viewed as _wrong_ and not just a faulty assumption on the part of C#/Win32.

**Comment by kennykerr** [comment]
This comes from the win32metadata. But yes, I agree that this isn't terribly reliable as `HANDLE` is used for all kinds of things that don't map to `CloseHandle`. That's why Rust ignores the `RAIIFree` attribute.

![image](https://user-images.githubusercontent.com/9845234/131411729-1bb6064f-7152-4fbb-8636-f01be7a64253.png)

**Comment by marler8997** [comment]
I would have expected the `RAIIFree` attribute to be applied to return types and out parameters rather than the types themselves.

**Comment by mikebattista** [comment]
Yes it's encoded in the metadata as Kenny pointed out with the RAIIFree attribute. We have more specific HANDLE structs for scenarios that require a more domain-specific Close/Free function.

In this case with LoadImage, you need to use 1 of 3 different "Free" functions based on the combination of parameters supplied, and CloseHandle isn't one of them.

**Comment by riverar** [comment]
OK, sounds like we want to codify that all Win32 HANDLEs be freeable via `CloseHandle` and any API that produces HANDLEs that do not comply get mutated to return a different type appropriate to that function. 

I can get behind that now, thanks for your patience while I wrapped my head around this.

**Comment by mikebattista** [comment]
Steve, could you take a look at this please? I ran into an issue remapping the return type here in remap.rsp since I think the way HANDLE is scraped is taking precedence over the manual remap.

**Comment by jevansaks** [comment]
Re-activating since the fix for this was lost in https://github.com/microsoft/win32metadata/commit/d1fa00a97f82f2a068241d5ace0a37665a63397e#diff-12f77aafef2c6ee2fa3a7c437254fc543c6e4246ffee487af8ae24db6d8d5bf0, which was fixing #1575.
